File: apps/ui/modules/happier-crypto-worker/ios/HappierCryptoWorkerModule.swift
import ExpoModulesCore
import Foundation

public final class HappierCryptoWorkerModule: Module {
  private static let workerQueue = DispatchQueue(
    label: "dev.happier.crypto-worker",
    qos: .userInitiated,
    attributes: .concurrent
  )

  public func definition() -> ModuleDefinition {
    Name("HappierCryptoWorker")

    AsyncFunction("getCapabilities") {
      HappierCryptoWorker.capabilities()
    }.runOnQueue(Self.workerQueue)

    AsyncFunction("echoBatchForDiagnostics") { (values: [String]) in
      HappierCryptoWorker.echoBatchForDiagnostics(values)
    }.runOnQueue(Self.workerQueue)

    AsyncFunction("decryptDataKeyEnvelopeV1Batch") { (items: [[String: String]]) in
      HappierCryptoWorker.decryptDataKeyEnvelopeV1Batch(items)
    }.runOnQueue(Self.workerQueue)

    AsyncFunction("decryptSecretboxJsonBatch") { (items: [[String: String]]) in
      HappierCryptoWorker.decryptSecretboxJsonBatch(items)
    }.runOnQueue(Self.workerQueue)

    AsyncFunction("decryptAesGcmJsonBatch") { (items: [[String: String]]) in
      HappierCryptoWorker.decryptAesGcmJsonBatch(items)
    }.runOnQueue(Self.workerQueue)
  }
}

File: apps/ui/modules/happier-crypto-worker/android/src/main/java/dev/happier/cryptoworker/HappierCryptoWorkerModule.kt
package dev.happier.cryptoworker

import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class HappierCryptoWorkerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HappierCryptoWorker")

    AsyncFunction("getCapabilities") {
      return@AsyncFunction HappierCryptoWorker.capabilities()
    }.runOnQueue(Queues.DEFAULT)

    AsyncFunction("echoBatchForDiagnostics") { values: List<String> ->
      return@AsyncFunction HappierCryptoWorker.echoBatchForDiagnostics(values)
    }.runOnQueue(Queues.DEFAULT)

    AsyncFunction("decryptDataKeyEnvelopeV1Batch") { items: List<Map<String, String>> ->
      return@AsyncFunction HappierCryptoWorker.decryptDataKeyEnvelopeV1Batch(items)
    }.runOnQueue(Queues.DEFAULT)

    AsyncFunction("decryptSecretboxJsonBatch") { items: List<Map<String, String>> ->
      return@AsyncFunction HappierCryptoWorker.decryptSecretboxJsonBatch(items)
    }.runOnQueue(Queues.DEFAULT)

    AsyncFunction("decryptAesGcmJsonBatch") { items: List<Map<String, String>> ->
      return@AsyncFunction HappierCryptoWorker.decryptAesGcmJsonBatch(items)
    }.runOnQueue(Queues.DEFAULT)
  }
}
